### ISSUE TITLE
[Bug] sgl-deep-gemm 0.2.0: SM90 weight-scale transform returns a non-owning alias (follow-up to #32188)

### ISSUE THREAD

**Issue by YingchaoX** [sgl-project/sglang#39684]
### Checklist

- [x] I searched related issues and PRs. #32188 fixes the SGLang activation-scale wrapper; this reports a remaining library API path. See also the closed #32097.
- [x] The bug persists in the latest **sgl-deep-gemm PyPI release, 0.2.0**, checked on 2026-09-16. The reproducer does not import SGLang or vLLM.
- [x] Environment information and a minimal reproducer are included.
- [x] This is a bug report written in English.

### Describe the bug

On SM90, `deep_gemm.transform_sf_into_required_layout` can return a tensor aliasing the input scale without retaining the original allocation's owner. Deleting the input allows the PyTorch caching allocator to reuse its storage while the returned tensor is still live. The output then changes to the contents of an unrelated allocation.

Reproduced with the published x86_64 wheels for **sgl-deep-gemm 0.1.7 and 0.2.0**, using PyTorch 2.13.0+cu130 and apache-tvm-ffi 0.1.11 on NVIDIA H20-3e (SM90). Version 0.2.0 was installed into a separate `--target` directory and selected via `PYTHONPATH`; both its distribution version and imported module path were checked. The original shared installation was not modified.

This is a follow-up to **Bug 1 in #32188**, not a new discovery of the general ownership defect. That PR preserves ownership in SGLang's `get_mn_major_tma_aligned_tensor` wrapper and explicitly leaves the library-level fix as follow-up. Here a downstream caller uses `transform_sf_into_required_layout` directly for **weight scales**, so it does not receive that protection. #32097 also identifies this broader API impact.

I am filing here because Issues are disabled on `sgl-project/DeepGEMM`; please redirect this report if another tracking location is preferred.

### Reproduction

Save the following as `repro.py`. Select one idle SM90 GPU. No model, distributed setup, or GEMM execution is required.

```python
"""Run on one idle SM90 GPU; no model or inference framework is required."""
import gc
import importlib.metadata
import json
import sys

import torch
import deep_gemm

print(json.dumps({'sgl-deep-gemm': importlib.metadata.version('sgl-deep-gemm'),
                  'apache-tvm-ffi': importlib.metadata.version('apache-tvm-ffi'),
                  'torch': torch.__version__, 'module_file': deep_gemm.__file__}), flush=True)

clone_output = '--clone' in sys.argv
sf = torch.full((1, 128, 40), 0.0005, device='cuda', dtype=torch.float32)
source_ptr = sf.data_ptr()
out = deep_gemm.transform_sf_into_required_layout(
    sf, 16384, 5120, (1, 128, 128),
    num_groups=1, is_sfa=False, disable_ue8m0_cast=True,
)
if not isinstance(out, torch.Tensor):
    out = torch.from_dlpack(out)
if clone_output:
    out = out.clone()  # Diagnostic control: original owner is still alive here.
torch.cuda.synchronize()
before_nan = out.isnan().sum().item()
aliases_input = out.data_ptr() == source_ptr
del sf
gc.collect()
# Simulate the small allocations made by subsequent model loading/forward work.
buffers = [torch.full((1, 128, 40), float('nan'), device='cuda') for _ in range(20)]
torch.cuda.synchronize()
result = {
    'clone': clone_output,
    'aliases_input': aliases_input,
    'source_pointer_reused': any(x.data_ptr() == source_ptr for x in buffers),
    'before_nan': before_nan,
    'after_nan': out.isnan().sum().item(),
    'numel': out.numel(),
}
print(json.dumps(result), flush=True)
assert torch.isfinite(out).all().item(), 'Live output aliases recycled scale storage'
assert torch.allclose(out, torch.full_like(out, 0.0005))

```

```bash
CUDA_VISIBLE_DEVICES=0 uv run --no-project --python /path/to/venv/bin/python python repro.py
CUDA_VISIBLE_DEVICES=0 uv run --no-project --python /path/to/venv/bin/python python repro.py --clone
```

The environment must contain torch, apache-tvm-ffi 0.1.11, and sgl-deep-gemm 0.2.0. Tested with the default PyTorch caching allocator.

Observed on both 0.1.7 and 0.2.0, without cloning:

```json
{"clone": false, "aliases_input": true, "source_pointer_reused": true, "before_nan": 0, "after_nan": 5120, "numel": 5120}
```

The finite-output assertion fails. On 0.2.0 with `--clone`:

```json
{"clone": true, "aliases_input": false, "source_pointer_reused": true, "before_nan": 0, "after_nan": 0, "numel": 5120}
```

Both assertions pass. The NaN-filled buffers are a deliberate overwrite sentinel: they demonstrate allocator reuse, not a claim that the original model uses this exact fill operation. Pointer reuse depends on allocator behavior; `source_pointer_reused` makes that condition visible.

**Expected:** the returned tensor remains valid and unchanged after the caller releases its reference to the input, whether the transform copies data or returns an alias.

### Downstream impact and diagnosis

The issue was encountered in vLLM 0.29.0 with a Qwen3.8-27B-FP8 checkpoint. With external sgl-deep-gemm 0.1.7, the first `in_proj_qkvz` layer had finite inputs and weights, but 2040/5120 weight-scale values were NaN. Eventually all logits became NaN and generation degenerated into repeated tokens.

Controlled eager-mode checks on that original environment:

| Configuration | Result |
|---|---|
| External 0.1.7, original scale transform | First request fails the finite check |
| vLLM-bundled DeepGEMM | 6/6 short correctness probes pass |
| External 0.1.7, clone transformed scales while the original input is alive | 6/6 probes pass; DeepGEMM and E8M0 remain enabled |

These are short correctness checks, not full model-quality or performance validation. The full model was not rerun with 0.2.0; the standalone API reproducer above was.

The v0.1.7 source path is consistent with the observed lifetime failure:

- `csrc/utils/torch_compat.hpp::convert_to_torch_tensor` wraps `TensorView`'s raw pointer with `torch::from_blob` without retaining its owner.
- The SM90 FP32 `(128, 128)` weight-scale branch in `csrc/apis/layout.hpp` returns the unchanged tensor.
- `csrc/tvm_ffi_api.cpp::dg_transform_sf_into_required_layout` exports that non-owning view through DLPack.

Nsight Systems 2026.5.1 captured the original standalone lifetime reproducer: it contains only allocation-fill/check-related CUDA kernels, with no GEMM kernel. The direct evidence for corruption is pointer reuse and tensor values, not the profiler alone.

Could the library ensure that no-op scale-transform return values retain the input owner (or return independently owned storage), with a regression covering input deletion and allocator reuse? The clone above is a diagnostic control, not a proposed requirement to copy every transform.

### Environment

Collected from the original environment with `uv run --no-project --python /path/to/venv/bin/python python -m sglang.check_env`. For the isolated latest-release repro, only sgl-deep-gemm was overlaid to **0.2.0**; apache-tvm-ffi is **0.1.11**. Only one GPU was used.

<details>
<summary>sglang.check_env output</summary>

```text
Python: 3.12.13 (main, Mar  4 2026, 09:23:07) [GCC 11.4.0]
CUDA available: True
GPU 0,1,2,3,4,5,6,7: NVIDIA H20-3e
GPU 0,1,2,3,4,5,6,7 Compute Capability: 9.0
CUDA_HOME: /usr/local/cuda
NVCC: Cuda compilation tools, release 13.0, V13.0.88
CUDA Driver Version: 580.95.05
PyTorch: 2.13.0+cu130
sglang: 0.5.19
sglang-kernel: 0.4.6.post1
flashinfer_python: 0.6.18
flashinfer_cubin: Module Not Found
flashinfer_jit_cache: Module Not Found
triton: 3.7.1
transformers: 5.12.1
torchao: Module Not Found
numpy: 2.3.5
aiohttp: 3.14.3
fastapi: 0.136.3
huggingface_hub: 1.30.0
interegular: 0.3.3
modelscope: 1.40.0
orjson: 3.12.0
outlines: 0.1.11
packaging: 26.3
psutil: 7.2.2
pydantic: 2.14.0b1
python-multipart: 0.0.32
pyzmq: 27.2.0
uvicorn: 0.52.4
uvloop: 0.22.1
vllm: 0.29.0
xgrammar: 0.2.1
openai: 2.6.1
tiktoken: 0.14.0
anthropic: 1.4.0
litellm: Module Not Found
torchcodec: 0.15.0
NVIDIA Topology: 
	GPU0	GPU1	GPU2	GPU3	GPU4	GPU5	GPU6	GPU7	NIC0	NIC1	NIC2	NIC3	CPU Affinity	NUMA Affinity	GPU NUMA ID
GPU0	 X 	NV18	NV18	NV18	NV18	NV18	NV18	NV18	PIX	PHB	SYS	SYS	0-91	0		N/A
GPU1	NV18	 X 	NV18	NV18	NV18	NV18	NV18	NV18	PXB	PHB	SYS	SYS	0-91	0		N/A
GPU2	NV18	NV18	 X 	NV18	NV18	NV18	NV18	NV18	PHB	PIX	SYS	SYS	0-91	0		N/A
GPU3	NV18	NV18	NV18	 X 	NV18	NV18	NV18	NV18	PHB	PXB	SYS	SYS	0-91	0		N/A
GPU4	NV18	NV18	NV18	NV18	 X 	NV18	NV18	NV18	SYS	SYS	PIX	PHB	92-183	1		N/A
GPU5	NV18	NV18	NV18	NV18	NV18	 X 	NV18	NV18	SYS	SYS	PXB	PHB	92-183	1		N/A
GPU6	NV18	NV18	NV18	NV18	NV18	NV18	 X 	NV18	SYS	SYS	PHB	PIX	92-183	1		N/A
GPU7	NV18	NV18	NV18	NV18	NV18	NV18	NV18	 X 	SYS	SYS	PHB	PXB	92-183	1		N/A
NIC0	PIX	PXB	PHB	PHB	SYS	SYS	SYS	SYS	 X 	PHB	SYS	SYS				
NIC1	PHB	PHB	PIX	PXB	SYS	SYS	SYS	SYS	PHB	 X 	SYS	SYS				
NIC2	SYS	SYS	SYS	SYS	PIX	PXB	PHB	PHB	SYS	SYS	 X 	PHB				
NIC3	SYS	SYS	SYS	SYS	PHB	PHB	PIX	PXB	SYS	SYS	PHB	 X 				

Legend:

  X    = Self
  SYS  = Connection traversing PCIe as well as the SMP interconnect between NUMA nodes (e.g., QPI/UPI)
  NODE = Connection traversing PCIe as well as the interconnect between PCIe Host Bridges within a NUMA node
  PHB  = Connection traversing PCIe as well as a PCIe Host Bridge (typically the CPU)
  PXB  = Connection traversing multiple PCIe bridges (without traversing the PCIe Host Bridge)
  PIX  = Connection traversing at most a single PCIe bridge
  NV#  = Connection traversing a bonded set of # NVLinks

NIC Legend:

  NIC0: mlx5_0
  NIC1: mlx5_1
  NIC2: mlx5_2
  NIC3: mlx5_3


Hypervisor vendor:: KVM
ulimit soft: 1024

```
</details>

**Comment by shtjww** [comment]
I'd like to pick this up. Plan: reproduce on H20 (SM90) with sgl-deep-gemm 0.2.0, then fix the ownership at the library level in DeepGEMM (`transform_sf_into_required_layout` should return a tensor that owns/retains the storage), plus a regression test that frees the input and verifies the output. One question: do you prefer the fix in `sgl-project/DeepGEMM` directly, or a defensive clone in SGLang's wrapper for now? Happy to go either way.

**Comment by KashmirAwana** [comment]
It looks like the issue is about `transform_sf_into_required_layout` returning a tensor that doesn't properly manage memory ownership, which can lead to unexpected changes in the output. I'd check the memory management logic in that function to see why it's not retaining ownership.

**Comment by shtjww** [comment]
Update: reproduced on H20 (SM90) with sgl-deep-gemm 0.2.0 — the standalone reproducer matches the reported numbers exactly (`aliases_input: true, after_nan: 5120`).

Root cause confirmed at the binding layer: `convert_to_torch_tensor()` wraps the input with a non-owning `torch::from_blob` view, and the `check_sf_layout` pass-through returns that view itself, so the DLPack export dangles once the caller frees the input.

Fix submitted at the library level: sgl-project/DeepGEMM#91 — adds an own-or-clone guard on all five scale-transform exits (zero-copy preserved on non-aliasing paths), plus a regression test that fails on the published 0.2.0 wheel and passes with the fix.

**Comment by shtjww** [comment]
Update: two PRs are ready — the sglang-side wrapper guard at #39977 (lint clean, synced with main, just needs the `run-ci` label) and the library-side root fix at sgl-project/DeepGEMM#91 (ownership guards on all 5 scale-transform exit points + regression tests). Both verified against the reproducer on H20. Happy to adjust based on maintainer feedback.
